common/src/jni/main/cpp/NativeCrypto.cpp

```diff
@@ -4311,6 +4311,202 @@ static void NativeCrypto_ASN1_TIME_to_Calendar(JNIEnv* env, jclass, jlong asn1Ti
                         sec);
 }
 
+// A CbsHandle is a structure used to manage resources allocated by asn1_read-*
+// functions so that they can be freed properly when finished.  This struct owns
+// all objects pointed to by its members.
+struct CbsHandle {
+    // A pointer to the CBS.
+    std::unique_ptr<CBS> cbs;
+    // A pointer to the data held by the CBS.  If the data held by the CBS
+    // is owned by a different CbsHandle, data will be null.
+    std::unique_ptr<unsigned char[]> data;
+};
+
+static jlong NativeCrypto_asn1_read_init(JNIEnv* env, jclass, jbyteArray data) {
+    JNI_TRACE("asn1_read_init(%p)", data);
+
+    ScopedByteArrayRO bytes(env, data);
+    if (bytes.get() == nullptr) {
+        Errors::throwIOException(env, "Error reading ASN.1 encoding");
+        return 0;
+    }
+
+    std::unique_ptr<CbsHandle> cbs(new CbsHandle());
+    cbs->data.reset(new unsigned char[bytes.size()]);
+    memcpy(cbs->data.get(), bytes.get(), bytes.size());
+
+    cbs->cbs.reset(new CBS());
+    CBS_init(cbs->cbs.get(), cbs->data.get(), bytes.size());
+    JNI_TRACE("asn1_read_init(%p) => %p", data, cbs.get());
+    return reinterpret_cast<uintptr_t>(cbs.release());
+}
+
+static jlong NativeCrypto_asn1_read_sequence(JNIEnv* env, jclass, jlong cbsRef) {
+    CbsHandle* cbs = reinterpret_cast<CbsHandle*>(static_cast<uintptr_t>(cbsRef));
+    JNI_TRACE("asn1_read_sequence(%p)", cbs);
+
+    std::unique_ptr<CbsHandle> seq(new CbsHandle());
+    seq->cbs.reset(new CBS());
+    if (!CBS_get_asn1(cbs->cbs.get(), seq->cbs.get(), CBS_ASN1_SEQUENCE)) {
+        Errors::throwIOException(env, "Error reading ASN.1 encoding");
+        return 0;
+    }
+    JNI_TRACE("asn1_read_sequence(%p) => %p", cbs, seq.get());
+    return reinterpret_cast<uintptr_t>(seq.release());
+}
+
+static jbyteArray NativeCrypto_asn1_read_octetstring(JNIEnv* env, jclass, jlong cbsRef) {
+    CbsHandle* cbs = reinterpret_cast<CbsHandle*>(static_cast<uintptr_t>(cbsRef));
+    JNI_TRACE("asn1_read_octetstring(%p)", cbs);
+
+    std::unique_ptr<CBS> str(new CBS());
+    if (!CBS_get_asn1(cbs->cbs.get(), str.get(), CBS_ASN1_OCTETSTRING)) {
+        Errors::throwIOException(env, "Error reading ASN.1 encoding");
+        return 0;
+    }
+    ScopedLocalRef<jbyteArray> out(env, env->NewByteArray(static_cast<jsize>(CBS_len(str.get()))));
+    if (out.get() == nullptr) {
+        Errors::throwIOException(env, "Error reading ASN.1 encoding");
+        return 0;
+    }
+    ScopedByteArrayRW outBytes(env, out.get());
+    if (outBytes.get() == nullptr) {
+        Errors::throwIOException(env, "Error reading ASN.1 encoding");
+        return 0;
+    }
+    memcpy(outBytes.get(), CBS_data(str.get()), CBS_len(str.get()));
+    JNI_TRACE("asn1_read_octetstring(%p) => %p", cbs, out.get());
+    return out.release();
+}
+
+static jlong NativeCrypto_asn1_read_integer(JNIEnv* env, jclass, jlong cbsRef) {
```

Comment:
I'd suggest naming this .._asn1_read_uint64. It's quite easy to misunderstand what "integer" is. Moreover, it should be clear that we're reading an unsigned value at the native layer, that is then getting cast to jlong...